Fixes issue #6898 FileWatcherService locks UI Thread on linux. In

mkrueger opened the pull request:
conjunction with the mono patch:
https://github.com/mono/mono/pull/12432

I think the mono patch is the real fix here but I optimized the
UpdateWatchers.

mkrueger committed:
Fixes issue #6898 FileWatcherService locks UI Thread on linux. In
conjunction with the mono patch:
https://github.com/mono/mono/pull/12432

I think the mono patch is the real fix here but I optimized the
UpdateWatchers.

Therzok commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects/FileWatcherService.cs:
```diff
@@ -78,48 +78,47 @@ static Task UpdateWatchersAsync ()
 				// First remove the watchers, so we don't spin too many threads.
-				foreach (var directory in toRemove) {
-					RemoveWatcher_NoLock (directory);
+				if (newWatchers.Count == 0) {
```
Do we ever hit this case as a fast path?

Therzok replied:
That would mean that we cleared all the filewatchers, which I don't think will happen often. I would go for simplicity of code here and merge the loop of this and the one below.

`newWatchers.Contains (directory)` in the second loop below will exit early as the collection has no elements.

mkrueger replied:
It was in the original code :), therefore I decided not to remove it. But y doesn't make much difference here.

Therzok replied:
We can have up to 8 watchers, due to the algorithm.

mkrueger committed:
Update FileWatcherService.cs

Therzok commented:
Okay, I remember why I added the extra list. See test failures:
```
2019-01-16T08:08:27.6125640Z    System.InvalidOperationException : Collection was modified; enumeration operation may not execute.
2019-01-16T08:08:27.6126540Z   at System.Collections.Generic.Dictionary`2+Enumerator[TKey,TValue].MoveNext () [0x00090] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-08/external/bockbuild/builds/mono-x64/external/corefx/src/Common/src/CoreLib/System/Collections/Generic/Dictionary.cs:1202 
2019-01-16T08:08:27.6126750Z   at MonoDevelop.Projects.FileWatcherService.UpdateWatchers (System.Threading.CancellationToken token) [0x00100] in /Users/vsts/agent/2.144.0/work/1/s/monodevelop/main/src/core/MonoDevelop.Core/MonoDevelop.Projects/FileWatcherService.cs:108 
2019-01-16T08:08:27.6126930Z   at MonoDevelop.Projects.FileWatcherService+<>c__DisplayClass8_0.<UpdateWatchersAsync>b__0 () [0x00000] in /Users/vsts/agent/2.144.0/work/1/s/monodevelop/main/src/core/MonoDevelop.Core/MonoDevelop.Projects/FileWatcherService.cs:79 
2019-01-16T08:08:27.6127780Z   at System.Threading.Tasks.Task.InnerInvoke () [0x0000f] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-08/external/bockbuild/builds/mono-x64/external/corert/src/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs:2500 
2019-01-16T08:08:27.6128670Z   at System.Threading.Tasks.Task.Execute () [0x00000] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-08/external/bockbuild/builds/mono-x64/external/corert/src/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs:2343 
2019-01-16T08:08:27.6129340Z --- End of stack trace from previous location where exception was thrown ---
```

Also, it looks like the line endings were changed.

mkrueger committed:
[Core] Fixed unit tests.

The toRemove list can easily be converted to a static field as well -
only used in a static method inside a lock without a way of recursion.

Therzok commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects/FileWatcherService.cs:
```diff
@@ -78,8 +78,9 @@ static Task UpdateWatchersAsync ()
-		static HashSet<FilePath> newWatchers = new HashSet<FilePath>();
-
+		static HashSet<FilePath> newWatchers = new HashSet<FilePath>();
+		static List<FilePath> toRemove = new List<FilePath> ();
```
toRemove is never cleared

Therzok reviewed (approved):
It is, misread.
